src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java

```diff
@@ -192,26 +189,27 @@ public StatementRestrictions(ClientState state,
          */
         for (Relation relation : whereClause.relations)
         {
-            if ((relation.isContains() || relation.isContainsKey()) && (type.isUpdate() || type.isDelete()))
+
+            Operator operator = relation.operator();
+            if (operator.requiresFilteringOrIndexingFor(ColumnMetadata.Kind.CLUSTERING) && (type.isUpdate() || type.isDelete()))
             {
-                throw invalidRequest("Cannot use %s with %s", type, relation.operator());
+                throw invalidRequest("Cannot use %s with %s", type, operator);
             }
 
-            if (relation.operator() == Operator.IS_NOT)
+            if (operator == Operator.IS_NOT)
             {
                 if (!forView)
                     throw new InvalidRequestException("Unsupported restriction: " + relation);
 
-                this.notNullColumns.addAll(relation.toRestriction(table, boundNames).getColumnDefs());
+                this.notNullColumns.addAll(relation.toRestriction(table, boundNames).columns());
             }
-            else if (relation.isLIKE())
+            else if (operator.requiresIndexing())
             {
                 Restriction restriction = relation.toRestriction(table, boundNames);
 
                 if (!type.allowUseOfSecondaryIndices() || !restriction.hasSupportingIndex(indexRegistry))
```

Comment:
nit: Could just use `indexRegistry.listIndexes()` and then IR doesn't have to be `Iterable`. Also, what happens if `indexRegistry` is `null`?

Reply:
At this places the code make sure the IndexRegistry is not null.